### PR TITLE
[CPP Graph] AMX-BF16 MHA with KV update

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: Feature
API changed: `--memory-f16` and `--memory-auto` added along with the old `--memory-f32` flag

## Description
AMX BF16 MHA with customized KV-cache update stage to maintain the customized memory format for best performance of MHA.
- The kv-update operation can be optimized if necessary; currently reference implementations are used
- S8 MHA are also added for future use, the corresponding update method will be adding in future changes
- BF16 MHA with transposed K-tensor add; integration work will be in a future PR
- FP16MHA add; integration work will be in a future PR

## Expected Behavior & Potential Risk
Note: As this PR updates jblas code, **models need to be re-quantized**.

## How has this PR been tested?
Runing `run_llama` & `run_gptj`

### Model Preparation
- `python scripts/convert_gptj.py  --outtype f32 --outfile gpt-j-6B.bin gpt-j-6B`
- `./bin/quant_gptj --model_file gpt-j-6B.bin --out_file gptj-q4j-int8-fp32-b128.bin --bits 4 --nthread 20 --alg sym --block_size 128 --scale_dtype fp32 --compute_type int8`
- Similar commands for `llama_7b`

### Inference result

- <details>
  <summary>
  GPTJ: AMX-BF16 MHA with KV update <b>enabled</b> - <code>(yi) [dingyi@mlp-spr-03 build]$ ninja && numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gptj-pr165-q4j-int8-fp32-b128.bin --seed 1234 -t 56 -b 512 -c 512 -n 256 -p "Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun."</code>
  </summary>

  > Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. Her mom had this really cool job at the local library, so they could afford all of these things for their girl. They also went on lots of trips together: Disneyland, SeaWorld, the Space Center, just about any city around them. The mother loved her work, and the daughter loved going places. It was a magical time in their lives, that made them both happy.
  >
  > So then mom got promoted to become a librarian in a small town in California. This meant that she had to move from Sacramento to Bakersfield - quite a drive - and they couldn't afford moving trucks, so the mother walked! They were going into this new city, and their daughter was excited - this was going to be fun!
  >
  > It turned out to be not much fun at all - the girl hated it. She didn't like her school or class mates. She didn't even like her teacher very much. And she missed her mom terribly, who was gone for weeks and weeks at a time, only coming home in between visits. Her parents were stressed about money, too, because their job moved them from Bakersfield to a bigger city, where they had to get jobs so they could afford rent and food and clothing.
  >
  > Sometimes it seemed like they were just
</details>

- <details>
  <summary>
  GPTJ: AMX-BF16 MHA with KV update <b>disabled</b> - <code>(yi) [dingyi@mlp-spr-03 build]$ ninja && numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gptj-pr165-q4j-int8-fp32-b128.bin --memory-f16 --seed 1234 -t 56 -b 512 -c 512 -n 256 -p "Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun."</code>
  </summary>

  > Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. Her friends would help her with this because they knew that she was curious about everything.
  > 
  > One day while talking to her mom on the phone, she asked her "Mommy, why do people cry?" Hearing her child say so made her mother laugh a little. She told her daughter that sometimes people will have tears in their eyes or eyes filled with water because they're happy or sad. But, usually people do not go around crying all day, which is where the girl's question came from.
  > 
  > "Well," she was told on the other end of the phone, "you mustn't cry for too long either. Otherwise you'll get tired." Her mother explained further that sometimes it happens because we're so happy about something that we can't stop laughing or crying and it won't stop for quite a while.
  > 
  > So, her mom told her to ask some questions. She was sure she'd be able to think of a few good ones. "Can you give me an example?" was what she asked at the beginning. But, as soon as she said this, she found herself thinking about a bunch of different scenarios and trying to figure out which one would make sense for the little girl.
  > 
  > Her mom's advice stuck in her mind,

  Note: this generation is the same as that on the main branch.
</details>

### Inference latency

- <details>
  <summary>
  GPTJ: AMX-BF16 MHA with KV update <b>enabled</b> (default on SPR with GCC13) - <code>(yi) [dingyi@mlp-spr-03 build]$ ninja && ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gptj-pr165-q4j-int8-fp32-b128.bin --seed 1234 -t 56 -b 512 -c 512 -n 256 -p "Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun."</code>
  </summary>

  First token
  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   2.682 ms
  perf_total_per_op_us[                     MUL] =   0.882 ms
  perf_total_per_op_us[                    NORM] =   3.717 ms
  perf_total_per_op_us[                 MUL_MAT] =   3.837 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =   1.883 ms
  perf_total_per_op_us[                 RESHAPE] =   0.267 ms
  perf_total_per_op_us[                    VIEW] =   0.754 ms
  perf_total_per_op_us[                 PERMUTE] =   0.093 ms
  perf_total_per_op_us[                GET_ROWS] =   0.286 ms
  perf_total_per_op_us[                    ROPE] =   2.093 ms
  perf_total_per_op_us[                 MUL_QKV] =  12.131 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] =  23.225 ms
  perf_total_per_op_us[              FLASH_ATTN] =   1.607 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  22.759 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  Last token
  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.451 ms
  perf_total_per_op_us[                     MUL] =   0.172 ms
  perf_total_per_op_us[                    NORM] =   1.060 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =   0.609 ms
  perf_total_per_op_us[                 RESHAPE] =   0.274 ms
  perf_total_per_op_us[                    VIEW] =   0.751 ms
  perf_total_per_op_us[                 PERMUTE] =   0.124 ms
  perf_total_per_op_us[                GET_ROWS] =   0.023 ms
  perf_total_per_op_us[                    ROPE] =   1.570 ms
  perf_total_per_op_us[                 MUL_QKV] =   5.120 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] =  12.097 ms
  perf_total_per_op_us[              FLASH_ATTN] =   1.841 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   1.385 ms
  perf_total_per_op_us[           INNER PRODUCT] =   2.709 ms
  ========================================
  ```
  Prediction latency
  ```
  ========== eval time log of each prediction ==========
  prediction   0, time: 79.93ms
  prediction   1, time: 32.00ms
  prediction   2, time: 30.42ms
  prediction   3, time: 30.79ms
  prediction   4, time: 30.74ms
  prediction   5, time: 30.57ms
  prediction   6, time: 30.76ms
  prediction   7, time: 30.09ms
  prediction   8, time: 30.78ms
  prediction   9, time: 30.96ms
  ...
  prediction 250, time: 31.67ms
  prediction 251, time: 31.65ms
  prediction 252, time: 30.93ms
  prediction 253, time: 31.59ms
  prediction 254, time: 31.10ms
  prediction 255, time: 31.01ms
  ```
</details>

- <details>
  <summary>
  GPTJ: AMX-BF16 MHA with KV update <b>disabled</b> - <code>(yi) [dingyi@mlp-spr-03 build]$ ninja && ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gptj-pr165-q4j-int8-fp32-b128.bin --memory-f16 --seed 1234 -t 56 -b 512 -c 512 -n 256 -p "Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun."</code>
  </summary>

  First token
  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   3.907 ms
  perf_total_per_op_us[                     MUL] =   0.963 ms
  perf_total_per_op_us[                    NORM] =   8.956 ms
  perf_total_per_op_us[                 MUL_MAT] =  14.315 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =   2.015 ms
  perf_total_per_op_us[                   SCALE] =   0.538 ms
  perf_total_per_op_us[                     CPY] =  36.292 ms
  perf_total_per_op_us[                 RESHAPE] =   0.288 ms
  perf_total_per_op_us[                    VIEW] =   0.851 ms
  perf_total_per_op_us[                 PERMUTE] =   0.378 ms
  perf_total_per_op_us[                GET_ROWS] =   0.320 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.845 ms
  perf_total_per_op_us[                SOFT_MAX] =   0.924 ms
  perf_total_per_op_us[                    ROPE] =   2.894 ms
  perf_total_per_op_us[                 MUL_QKV] =   9.531 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] =  33.988 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  Last token
  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.420 ms
  perf_total_per_op_us[                     MUL] =   0.154 ms
  perf_total_per_op_us[                    NORM] =   1.053 ms
  perf_total_per_op_us[                 MUL_MAT] =   4.255 ms
  perf_total_per_op_us[       MUL_MAT_WITH_BIAS] =   0.624 ms
  perf_total_per_op_us[                   SCALE] =   0.114 ms
  perf_total_per_op_us[                     CPY] =   3.408 ms
  perf_total_per_op_us[                 RESHAPE] =   0.292 ms
  perf_total_per_op_us[                    VIEW] =   0.907 ms
  perf_total_per_op_us[                 PERMUTE] =   0.374 ms
  perf_total_per_op_us[                GET_ROWS] =   0.021 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.097 ms
  perf_total_per_op_us[                SOFT_MAX] =   0.788 ms
  perf_total_per_op_us[                    ROPE] =   1.603 ms
  perf_total_per_op_us[                 MUL_QKV] =   4.958 ms
  perf_total_per_op_us[            FFN_ADD_GeLU] =  12.166 ms
  perf_total_per_op_us[           INNER PRODUCT] =   2.689 ms
  ========================================
  ```
  Prediction latency
  ```
  ========== eval time log of each prediction ==========
  prediction   0, time: 121.24ms
  prediction   1, time: 39.97ms
  prediction   2, time: 38.27ms
  prediction   3, time: 38.43ms
  prediction   4, time: 38.17ms
  prediction   5, time: 38.09ms
  prediction   6, time: 38.10ms
  prediction   7, time: 38.19ms
  prediction   8, time: 38.79ms
  prediction   9, time: 37.90ms
  ...
  prediction 250, time: 38.81ms
  prediction 251, time: 38.08ms
  prediction 252, time: 38.89ms
  prediction 253, time: 37.40ms
  prediction 254, time: 38.84ms
  prediction 255, time: 37.94ms
  ```
</details>

- <details>
  <summary>
  GPTJ: AMX-BF16 MHA with KV update disabled <b>on the main branch</b> (but with jblas updated) - <code>(yi) [dingyi@mlp-spr-03 build]$ ninja && ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_gptj -m /home_local/dingyi/data/gptj-pr165-q4j-int8-fp32-b128.bin --seed 1234 -t 56 -b 512 -c 512 -n 256 -p "Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun."</code>
  </summary>

  First token
  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[             ADD] =   3.957 ms
  perf_total_per_op_us[             MUL] =   1.316 ms
  perf_total_per_op_us[            NORM] =  13.196 ms
  perf_total_per_op_us[         MUL_MAT] =  16.293 ms
  perf_total_per_op_us[MUL_MAT_WITH_BIAS] =   1.915 ms
  perf_total_per_op_us[           SCALE] =   0.557 ms
  perf_total_per_op_us[             CPY] =  35.847 ms
  perf_total_per_op_us[         RESHAPE] =   0.276 ms
  perf_total_per_op_us[            VIEW] =   0.841 ms
  perf_total_per_op_us[         PERMUTE] =   0.365 ms
  perf_total_per_op_us[        GET_ROWS] =   0.289 ms
  perf_total_per_op_us[   DIAG_MASK_INF] =   1.214 ms
  perf_total_per_op_us[        SOFT_MAX] =   1.238 ms
  perf_total_per_op_us[            ROPE] =   4.515 ms
  perf_total_per_op_us[         MUL_QKV] =  11.250 ms
  perf_total_per_op_us[    FFN_ADD_GeLU] =  30.236 ms
  perf_total_per_op_us[   INNER PRODUCT] =   0.000 ms
  ========================================
  ```
  Last token
  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[             ADD] =   0.422 ms
  perf_total_per_op_us[             MUL] =   0.159 ms
  perf_total_per_op_us[            NORM] =   1.100 ms
  perf_total_per_op_us[         MUL_MAT] =   4.340 ms
  perf_total_per_op_us[MUL_MAT_WITH_BIAS] =   0.589 ms
  perf_total_per_op_us[           SCALE] =   0.118 ms
  perf_total_per_op_us[             CPY] =   2.715 ms
  perf_total_per_op_us[         RESHAPE] =   0.291 ms
  perf_total_per_op_us[            VIEW] =   0.820 ms
  perf_total_per_op_us[         PERMUTE] =   0.369 ms
  perf_total_per_op_us[        GET_ROWS] =   0.008 ms
  perf_total_per_op_us[   DIAG_MASK_INF] =   0.098 ms
  perf_total_per_op_us[        SOFT_MAX] =   0.820 ms
  perf_total_per_op_us[            ROPE] =   1.652 ms
  perf_total_per_op_us[         MUL_QKV] =   4.936 ms
  perf_total_per_op_us[    FFN_ADD_GeLU] =  12.455 ms
  perf_total_per_op_us[   INNER PRODUCT] =   2.549 ms
  ========================================
  ```
  Prediction latency
  ```
  ========== eval time log of each prediction ==========
  prediction   0, time: 127.38ms
  prediction   1, time: 43.01ms
  prediction   2, time: 42.12ms
  prediction   3, time: 40.99ms
  prediction   4, time: 41.40ms
  prediction   5, time: 38.56ms
  prediction   6, time: 36.69ms
  prediction   7, time: 37.65ms
  prediction   8, time: 36.81ms
  prediction   9, time: 37.25ms
  ...
  prediction 250, time: 37.15ms
  prediction 251, time: 37.44ms
  prediction 252, time: 36.92ms
  prediction 253, time: 37.18ms
  prediction 254, time: 37.12ms
  prediction 255, time: 37.35ms

  ```
</details>

Note:  `run_gptj` tested on ICX and tested with q4_0

## Dependency Change?
N/A

### CODE REVIEW

File: intel_extension_for_transformers/backends/neural_engine/graph/models/llama/llama.cpp [hunk 1]
```diff
@@ -138,13 +168,19 @@ static bool llama_model_eval_internal(model_context& lctx, const model_token* to
       Kcur = ne_rope_inplace(
           ctx0, ne_reshape_3d(ctx0, ne_mul_mat(ctx0, model.layers[il].attn[1], cur), n_embd / n_head, n_head, N),
           n_past, n_rot, 0);
-      Vcur = ne_transpose(ctx0, ne_reshape_2d(ctx0, ne_mul_mat(ctx0, model.layers[il].attn[2], cur), n_embd, N));
+      if (!kv_mem_jblas) {
+        Vcur = ne_transpose(ctx0, ne_reshape_2d(ctx0, ne_mul_mat(ctx0, model.layers[il].attn[2], cur), n_embd, N));
+      } else {
+        Vcur = ne_rope_inplace(
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/backends/neural_engine/graph/models/llama/llama.cpp:171]: `Vcur` does not need `rope_embedding`

File: intel_extension_for_transformers/backends/neural_engine/graph/models/gptj/gptj.cpp [hunk 2]
```diff
@@ -131,7 +154,16 @@ static bool gptj_model_eval_internal(model_context& lctx, const model_token* tok
                                                       1 * N * n_embd * batch_size * ne_element_size(QKVcur)),
                                            n_embd / n_head, n_head, N, batch_size),
                              n_past, n_rot, 0);
-      Vcur = ne_view_1d(ctx0, QKVcur, N * n_embd * batch_size, 2 * N * n_embd * batch_size * ne_element_size(QKVcur));
+      if (!kv_mem_jblas) {
+        Vcur = ne_view_1d(ctx0, QKVcur, N * n_embd * batch_size, 2 * N * n_embd * batch_size * ne_element_size(QKVcur));
+      } else {
+        Vcur = ne_rope_inplace(ctx0,
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/backends/neural_engine/graph/models/gptj/gptj.cpp]: `Vcur` does not need `rope_embeddings`

**Review by zhentaoyu — APPROVED** [review]
The `kv_cache_update` functions in `JBLAS` may need to be enlarged in the future if requested to enable new features and new models. For example (as far as I can see), 1.`decode_prompt length` + `max_new_tokens` > `n_ctx`, the `kv_cache` need a cyclic shift. 2. There will be more cache reorder (or copy) between different `beam_size` in the `beam_search` functions.

File: intel_extension_for_transformers/backends/neural_engine/graph/models/gptj/gptj.cpp (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/backends/neural_engine/graph/models/gptj/gptj.cpp]: Thanks. It has been fixed in 90e51f6 and the performance data in the PR description has been updated accordingly.

File: intel_extension_for_transformers/backends/neural_engine/graph/models/llama/llama.cpp (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/backends/neural_engine/graph/models/llama/llama.cpp:171]: Thanks. It has been fixed in 90e51f6 and the performance data in the PR description has been updated accordingly.

File: intel_extension_for_transformers/llm/runtime/graph/jblas/jblas/jit_base.hpp [hunk 3]
```diff
@@ -227,6 +227,26 @@ class JitAmxbf16 : protected JitAmxtile {
   void cvt_fp32_bf16(const Xbyak::Ymm& _bf16, const Xbyak::Zmm& _fp32) { vcvtneps2bf16(_bf16, _fp32); }
 };
 
-class JitAmxint8 : protected JitAmxtile {};
+class JitAmxint8 : protected JitAmxtile {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/jblas/jblas/jit_base.hpp:230]: there are 2 copies of jblas, can you remove this one?

**Review by airMeng — CHANGES_REQUESTED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c [hunk 4]
```diff
@@ -8395,47 +8422,125 @@ static void ne_compute_forward_flash_attn_f32_f16_f16(const struct ne_compute_pa
...
+      .step_v_head_size = 1,
       .step_dst_bs = seq_cur * embedsize,
       .step_dst_head_num = headsize,
       .step_dst_sl = embedsize,
-      .tmp = tmp->data,
   };
   jblas_fusion_attn_fp32_fp16_fp16_fp32_forward(&args);
 }
 
+static void ne_compute_forward_flash_attn_reordered(const struct ne_compute_params* params, const struct ne_tensor* q,
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:8483]: can you move all functions you updated to another file? seems ne_layers.c is much larger than expected

**Review by airMeng — CHANGES_REQUESTED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/jblas/jblas/jit_base.hpp (hunk 3 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/jblas/jblas/jit_base.hpp:230]: I think @zhenwei-intel will do the migration for us as part of the work of project directory restructure. I'll rebase the branch then.

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c (hunk 4 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:8483]: Agree that the `ne_layers.c` is too large.

If we move the implementation to a new file (say "ne_layers_mha.c"), will we also move their declarations to a new header file, or keep then in `ne_layers.h`?

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:8483]: headers might not change

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:8496]: `ne_xxx` and `ne_compute_xxx` functions are put in `ne_layers.c`, while detailed functions for actual checking & computing are declared in `layers/mha_dense.h` and implemented in `layers/mha_dense.cpp`

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_utils.h [hunk 5]
```diff
@@ -29,6 +29,7 @@
 #define CompileAVX512F() (defined(__GNUC__) && (__GNUC__ >= 6))
 #define CompileAVX2() (defined(__GNUC__) && (__GNUC__ >= 5))
 #define CompileAMX() (defined(__GNUC__) && (__GNUC__ >= 11))
+#define CompileBF16Kernel() (defined(__GNUC__) && (__GNUC__ >= 10))
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_utils.h:32]: what's the differences between ```CompileBF16Kernel``` and ```CompileBF16```

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 6]
```diff
@@ -50,20 +54,79 @@ struct attn_fwd_args_t {
...
+
+  const auto x1 = _mm512_fmadd_ps(x, log2e, half);  // auto x1 = x * log2e + _mm512_set1_ps(.5f);
+  const auto z = _mm512_floor_ps(x1);
+  const auto f = _mm512_sub_ps(x1, z);  // auto f = x1 - z;
+
+  return poly_scale_2nd_ps(z, f, c0, c1, c2);
+}
+
+inline float mha_exp_ref(float x) {
+#if MHA_2ND_EXP
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:91]: seems always 1?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 7]
```diff
@@ -50,20 +54,79 @@ struct attn_fwd_args_t {
...
+
+#ifdef NOT_CURRENTLY_USED
+inline __m512 exp_2nd_ph(const __m512 z, const __m512 f, const __m512 c0, const __m512 c1, const __m512 c2) {
+  const auto y = _mm512_fmadd_ph(_mm512_fmadd_ph(f, c0, c1), f, c2);  // auto y = (f * c0 + c1) * f + c2;
+  const auto exp = _mm512_scalef_ph(y, z);
+  return exp;
+}
+
+inline __m512 exp_ph_0_1(const __m512 x) {
+  static const auto c0 = _mm512_castsi512_ph(_mm512_set1_epi16(fp16(0.240226507f).x));
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:112]: @yuchengliu1 please align these in future

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.h [hunk 8]
```diff
@@ -23,40 +24,131 @@ extern "C" {
...
+  float QK_scale;
+  bool is_causal;
+  int batch_size, head_num, head_size, sl_q, sl_kv;
+  ATTN_FWD_LAYOUT Q_layout, K_layout, V_layout, dst_layout;
+  int step_q_bs, step_q_head_num, step_q_sl;
+  int step_k_bs, step_k_head_num, step_k_sl, step_k_head_size;
+  int step_v_bs, step_v_head_num, step_v_sl, step_v_head_size;
+  int step_dst_bs, step_dst_head_num, step_dst_sl;
+} attn_fp16_fwd_args_t;
+void jblas_fusion_attn_fp16_forward(const attn_fp16_fwd_args_t* params);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.h:101]: should there be a function ```jblas_fusion_attn_fp16_support``` ?

**Review by airMeng — APPROVED** [review]

**Comment by airMeng** [comment]
@zhentaoyu @zhenwei-intel @intellinjun @Zhenzhong1 hi maintainers of each model, please be aware of this new API

cc @a32543254 @kevinintel

File: intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_utils.h (hunk 5 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_utils.h:32]: I'm not sure. See https://github.com/intel-sandbox/jblas/pull/23/files#diff-4f85ef22dde7dcf1f27a8db354d0c00740b2143e744a33d54aafe3fd874e1307R32.
(internal access required).

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.h (hunk 8 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.h:102]: Good point! Added in 7d0a469.

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 6 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:91]: `MHA_2ND_EXP`may be turned off for accuracy validation.

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 7 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:112]: consider move to special math functions file

**Review by luoyu-intel — APPROVED** [review]
Need to validate q4_0 model and run model on clx device

**Comment by DDEle** [comment]
> Need to validate q4_0 model and run model on clx device

Successfully run with q4_0 and run on icx-1

**Review by airMeng — APPROVED** [review]

**Pull request merged**